### PR TITLE
WIP: Add API developer docs

### PR DESCRIPTION
**Pull request by Kludex**
It's been months since issue #804 was created and since that date I've seen a lot of people interested on it. On there I also saw  @tiangolo 's wishes about someday using [mkautodoc](https://github.com/tomchristie/mkautodoc), but the framework is not maintained anymore. On the other hand, [mkdocstrings](https://github.com/pawamoy/mkdocstrings) is being under develop since that issue and now has achieved a much more mature state. Due to that fact, I'm opening this PR that has the following changes:

- Add `API Documentation` section on our documentation.
- Add `mkdocstrings` plugin on `mkdocs`.
- Add initial docstrings following [Google style](https://google.github.io/styleguide/pyguide.html).

Images:
image
image

Notes:

- Recommendations about how to document `Starlette` code are accepted.
- Copy/paste of Starlette code and FastAPI documentation itself will be used when possible.
- [Darglint](https://github.com/terrencepreilly/darglint) will be used with `-z full` or `-z long` flag. Accept suggestions.
- I haven't talked to @tiangolo about this, and I do not know his wishes about it. Please do not expect this to be merged, but if you like the idea, please add a :+1: here.

### CODE REVIEW

**Comment by ycd** [comment]
Awesome effort @Kludex, the preview looks amazing  :rocket: :tada:

Also, Issues related to this PR:

#1588
#845 
#804

**Comment by tiangolo** [comment]
Thanks for all the work on this! That's very cool! :rocket: 

I want to add a couple of things to FastAPI that I still have in the backlog, and that will change some of these APIs a bit. So I want to have some of those things a bit more stable before adding more stable/defined API docs.

The other thing I want to do is to make sure the docstrings look nice on VS Code and PyCharm, so I imagine I will end up leaning towards some type of Markdown... but I still don't know any of that yet, I have to experiment a bit :grimacing:

**Comment by Kludex** [comment]
@tiangolo Thank you for letting me know!

I've just found out this: https://pypi.org/project/pydoc-markdown/ (Idk if it's useful)

**Comment by tiangolo** [comment]
Thanks!

I think you can leave this PR open because you already did a lot of the work needed for the setup that I think will be useful (even if I change a bit the format or something). :bow: I probably will not merge it soon, but I can use it later as the starting point :nerd_face:

**Comment by Kludex** [comment]
Perfect! I'll open it again. Thank you! :) 

Just for future references: I didn't add [darglint](https://github.com/terrencepreilly/darglint) but I was planing to.

**Comment by Kludex** [comment]
@tiangolo I've just discovered that the package I used, `mkdocstrings` supports Markdown :sweat_smile:

File: pyproject.toml [hunk 1]
```diff
@@ -67,6 +67,7 @@ doc = [
     "mkdocs >=1.1.2,<2.0.0",
     "mkdocs-material >=6.1.4,<7.0.0",
     "markdown-include >=0.5.1,<0.6.0",
+    "mkdocstrings == 0.13.6",
```
Comment (pawamoy) [inline pyproject.toml:70]: There's a new version, 0.14.0, if you want to update your PR 🙂

**Comment by Mause** [comment]
@tiangolo at this point it's been two years since https://github.com/tiangolo/fastapi/issues/804 was raised, at this point I think it's better to just implement it, rather than wait for perfection, which may never come

**Comment by cyprienc** [comment]
Bumping this

File: pyproject.toml (hunk 1 above)
Comment (pawamoy) [reply pyproject.toml:70]: Latest is now 0.16.0 🙂

**Comment by priyamroy** [comment]
+1

**Comment by adamcunnington** [comment]
Those project is going to die if things as trivial as this can't be contributed below perfection :(

**Comment by marianhlavac** [comment]
Any update on this? This seemed to be very close to finishing. At this point I agree with "better this than perfect".

**Comment by WorkShoft** [comment]
This issue needs some actionable feedback. Not having an API reference is surely frustrating some users. Are there any specific details that need to be improved before merging it?

**Comment by adamcunnington** [comment]
Err. Yes. API docstrings. I think the need is quite clear.

**Comment by rabernat** [comment]
Please please please merge this 🙏 . This project needs API docs. This PR implements API docs.

File: fastapi/applications.py [hunk 2]
```diff
@@ -228,6 +229,32 @@ def add_api_route(
         ),
         name: Optional[str] = None,
     ) -> None:
+        """[summary]
```
Comment (bencwallace) [inline fastapi/applications.py:232]: Looks like the summary is missing here and below.

**Comment by leokhachatorians** [comment]
Is it a deliberate design choice to not include API documentation?

**Comment by rabernat** [comment]
Perhaps we could consider deploying these API docs to a standalone site, separate from the main fastapi repo. They would clearly have a lot of value for the community. @Kludex would you be open to that possibility?

**Comment by pydanny** [comment]
I think a [standalone site publishing the API docs](https://twitter.com/pydanny/status/1529153162896060417) is a fantastic idea. It means a faster development cycle and lifts some of the burden from maintainers. This is a common pattern across a lot of tools.

I'm really excited to see what gets built!

**Comment by pawamoy** [comment]
Something like this: https://pawamoy.github.io/fastapi-ref/ :stuck_out_tongue:

**Comment by pawamoy** [comment]
I've updated my non-official demo site: https://pawamoy.github.io/fastapi-ref/reference/.
It documents version 0.101.1. It doesn't make any effort to organize docs, and just dumps *everything* ordered alphabetically.

**Comment by Kludex** [comment]
- Closing in favor of https://github.com/tiangolo/fastapi/pull/10392

**Pull request closed without merging**